File: project_euler/02/021.py
#!/usr/bin/env python3
'''Copyright [2017] <dmnsn7@gmail.com>'''

N = 10000


def get_fac_sum(num_n):
    fac_sum, tmp_n = 1, num_n
    for i in range(2, tmp_n + 1):
        if i > tmp_n // i:
            break
        if tmp_n % i == 0:
            to_multi, base = 1, 1
            while tmp_n % i == 0:
                base *= i
                to_multi += base
                tmp_n //= i
            fac_sum *= to_multi

    if tmp_n != 1:
        fac_sum *= tmp_n + 1

    return fac_sum - num_n


def is_amic(num_n):
    return get_fac_sum(num_n) != num_n and get_fac_sum(
        get_fac_sum(num_n)) == num_n


def main():
    amic_sum = 0
    for i in range(2, N):
        amic_sum += i if is_amic(i) else 0

    print(amic_sum)


if __name__ == '__main__':
    main()

File: project_euler/02/021.go
// Copyright [2017] <dmnsn7@gmail.com>

package main

import "fmt"

const cn = 10000

func getFacSum(n int) int {
	facSum, tn := 1, n
	for i := 2; i <= tn/i; i++ {
		if tn%i == 0 {
			toMulti, base := 1, 1
			for tn%i == 0 {
				base *= i
				toMulti += base
				tn /= i
			}
			facSum *= toMulti
		}
	}

	if tn != 1 {
		facSum *= tn + 1
	}

	return facSum - n
}

func isAmic(n int) bool {
	return getFacSum(n) != n && getFacSum((getFacSum(n))) == n
}

func main() {
	amicSum := 0
	for i := 2; i < cn; i++ {
		if isAmic(i) {
			amicSum += i
		}
	}

	fmt.Printf("%d\n", amicSum)
}
